internal/manifest/manifest.go

```diff
@@ -140,40 +147,94 @@ func (m *Manifest) Validate() error {
-		return nil, fmt.Errorf("failed to convert TrustedMeasurement from manifest to byte slices: %w", err)
-	}
-
-	return &validate.Options{
-		Measurement: trustedMeasurement,
-		GuestPolicy: abi.SnpPolicy{
-			Debug: false,
-			SMT:   true,
-		},
-		VMPL: new(int), // VMPL0
-		MinimumTCB: kds.TCBParts{
-			BlSpl:    m.ReferenceValues.AKS.SNP.MinimumTCB.BootloaderVersion.UInt8(),
-			TeeSpl:   m.ReferenceValues.AKS.SNP.MinimumTCB.TEEVersion.UInt8(),
-			SnpSpl:   m.ReferenceValues.AKS.SNP.MinimumTCB.SNPVersion.UInt8(),
-			UcodeSpl: m.ReferenceValues.AKS.SNP.MinimumTCB.MicrocodeVersion.UInt8(),
-		},
-		MinimumLaunchTCB: kds.TCBParts{
-			BlSpl:    m.ReferenceValues.AKS.SNP.MinimumTCB.BootloaderVersion.UInt8(),
-			TeeSpl:   m.ReferenceValues.AKS.SNP.MinimumTCB.TEEVersion.UInt8(),
-			SnpSpl:   m.ReferenceValues.AKS.SNP.MinimumTCB.SNPVersion.UInt8(),
-			UcodeSpl: m.ReferenceValues.AKS.SNP.MinimumTCB.MicrocodeVersion.UInt8(),
-		},
-		PermitProvisionalFirmware: true,
-	}, nil
+
+	var out []*SNPValidateOptsGenerator
+	for _, refVal := range m.ReferenceValues {
+		trustedMeasurement, err := refVal.SNP.TrustedMeasurement.Bytes()
+		if err != nil {
+			return nil, fmt.Errorf("failed to convert TrustedMeasurement from manifest to byte slices: %w", err)
+		}
+
+		out = append(out, &SNPValidateOptsGenerator{
+			manifest: m,
+			opts: &validate.Options{
+				Measurement: trustedMeasurement,
+				GuestPolicy: abi.SnpPolicy{
+					Debug: false,
+					SMT:   true,
+				},
+				VMPL: new(int), // VMPL0
+				MinimumTCB: kds.TCBParts{
+					BlSpl:    refVal.SNP.MinimumTCB.BootloaderVersion.UInt8(),
+					TeeSpl:   refVal.SNP.MinimumTCB.TEEVersion.UInt8(),
+					SnpSpl:   refVal.SNP.MinimumTCB.SNPVersion.UInt8(),
+					UcodeSpl: refVal.SNP.MinimumTCB.MicrocodeVersion.UInt8(),
+				},
+				MinimumLaunchTCB: kds.TCBParts{
+					BlSpl:    refVal.SNP.MinimumTCB.BootloaderVersion.UInt8(),
+					TeeSpl:   refVal.SNP.MinimumTCB.TEEVersion.UInt8(),
+					SnpSpl:   refVal.SNP.MinimumTCB.SNPVersion.UInt8(),
+					UcodeSpl: refVal.SNP.MinimumTCB.MicrocodeVersion.UInt8(),
+				},
+				PermitProvisionalFirmware: true,
+			},
+		})
+	}
+
+	if len(out) == 0 {
+		return nil, errors.New("no AKS reference values found in manifest")
+	}
+
+	return out, nil
+}
+
+// WithReportHostData augments the validate options generator with
+// a check that verifies whether the policy hash in the report's
+// HOSTDATA is allowed by the manifest.
+func (g *SNPValidateOptsGenerator) WithReportHostData() *SNPValidateOptsGenerator {
+	g.extraChecks = append(g.extraChecks, func(report *sevsnp.Report) error {
+		hostData := NewHexString(report.HostData)
+		if _, ok := g.manifest.Policies[hostData]; !ok {
+			return fmt.Errorf("hostdata %s not found in manifest", hostData)
+		}
+		return nil
+	})
+	return g
+}
+
+// WithStaticHostData augments the validate options with
+// the given host data.
+func (g *SNPValidateOptsGenerator) WithStaticHostData(hostData []byte) *SNPValidateOptsGenerator {
```

Comment:
> This would allow us to get rid of SNPValidateOptsGenerator (and StaticValidateOptsGenerator) as a concept entirely.

I don't think so. Based on my current understanding, the reason for having these dynamic generator functions is that we need to have callbacks, as we need to verify values from the report against the manifest. As the report is only known at the actual validation time (which happens after the options need to be available), we don't have a way to pass it in advance when generating the options. Maybe I misunderstood your proposal though.

Reply:
I'm talking about `SNPValidateOptsGenerator.extraChecks`. This field is only ever used to verify that the host data matches one of the several values in the manifest. When the `SNPValidateOptsGenerator` is created the manifest is already known because every `SNPValidateOptsGenerator` is created through a method on the manifest, `Manifest.SNPValidateOpts()`. Currently there is no other implementation of `validateOptsGenerator.SNPValidateOpts` that does anything else.

Reply:
The problem is not the availability of the manifest, but of the report. This is not known in advance

Reply:
I don't understand. AFAICT we don't need to have the report in order to emit a set of options for go-sev-guest to work with.

Reply:
> We don't necessarily need to to generate the options, but I would strongly advocate for keeping it here, since the dynamic checking in regards to the report (i.e. `WithReportHostData`) essentially does the same thing as `WithStaticHostData`, but from a dynamic value that is only known at attestation time (i.e. the report)

Why is checking the values dynamically from a callback preferable?

> Hence we had and have support for the "callbacks" (quoting since it's actually called _before_ the other validation) integrated into the option generation. Where would we put the "callback" with your proposal?

I'm only talking about `SNPValidateOptsGenerator.extraChecks`. My claim is that we don't need it. If we want to allow multiple host-data values, we should emit multiple validation options for go-sev-guest.

Reply:
> Why is checking the values dynamically from a callback preferable?

Because it's not possible to know the values from the report in advance. Let's discuss this in a call